### ISSUE TITLE
Confused about target meaning in attributions for 1D output classifier

### ISSUE THREAD

**Issue by pedrob37** [pytorch/captum#475]
## ❓ Questions and Help

I've been using Captum for my classification networks, but upon review I'm not sure how the interpretability works in my case. My network is a binary classifier with a 1D output (between zero and one) that takes in only tabular data. I run attributions using most methods (feature ablation, deep lift, etc.). 

My questions are these: 

Isn't the target parameter crucial to specify? What happens when it is set to None (default behaviour), what class are the attributions being compared against? 

In my case I have set the target to zero, but the output is 1D so does this function differently from setting it to None? How does Captum interpret this; i.e. against what class are my attributions in this case and what do the positive and negative values mean here? Does it even make sense to use Captum with a 1D output classification network?

If I changed my network to have an output size of two (one for each class), then using a softmax to determine class, would setting the target to 0 correspond to attributions referring to the first class, and setting the target to one correspond to attributions referring to the first class?

**Comment by vivekmig** [comment]
Hi @pedrob37 , you can definitely use Captum with a 1D output for binary classification. In this case, you can interpret the attributions as positive attributions leading to a higher output score and negative attributions as contributing to a lower output score. Since in binary classification, the output score generally corresponds to a probability / score of class 1, this is essentially like attribution with respect to class 1.

In order to attribute to the other class (class 0), you could create a wrapper function, which just returns 1 - model_output (or any other score corresponding to class 0), and compute attributions with this function. For many methods, the returned attributions will simply be equal to negating the original attributions, which makes sense since the significance is essentially reversed with two classes.

Regarding the difference between setting target to None or 0, this mainly depends on the exact shape of your models output. If the model output is a 1D tensor (output shape is num_examples), then target should generally be None, and setting it to 0 should likely cause an error. If the model output is a 2D tensor with second dimension 1 (output shape is num_examples x 1), then target can be 0, corresponding to selecting the first column, or None (which defaults to the single value), both should be equivalent and have the same results. 

If the output is changed to 2D (num_examples x 2), then target of 0 or 1 would determine the class being attributed to, as you expected. 

Our FAQ has some more context on the target argument [here](https://captum.ai/docs/faq#how-do-i-set-the-target-parameter-to-an-attribution-method) and this [tutorial](https://captum.ai/tutorials/IMDB_TorchText_Interpret) on the IMDB also has an example using a model with a scalar output for binary classification. Hope this helps!

**Comment by pedrob37** [comment]
Thank you, this has certainly clarified a lot of things! As a follow up, how exactly would I go about creating a wrapped for my model to reverse the output in that way? Given that I create an instance of a Captum interpretability class on my model, and not explicitly on the output?

**Comment by vivekmig** [comment]
Hi @pedrob37 , you can do this by just creating a function that calls the model and performs any additional steps. For methods that don't use hooks (most methods other than DeepLift, GuidedBackprop, Deconvolution), you can provide this function directly to the attribution method. The discussion in [this issue](https://github.com/pytorch/captum/issues/444) has some examples of this and a work-around for the methods using hooks.

**Comment by NarineK** [comment]
@pedrob37, can we close this issue ?
